### ISSUE TITLE
Orchard 1.9 very poor content type deletion performance 

### ISSUE THREAD

**Issue by byhub** [OrchardCMS/Orchard#7828]
we have a CRM system where user is able to create his own datasets which are essentially a content type, a content part (which is unique to each content type), and a set of fields in that part. Everything works smoothly except for the cases where we need to remove user's datasets: content type and part are being deleted extremely slowly, up to several seconds. Is there a way to speed up content type removal?

**Comment by Hazzamanic** [comment]
Several seconds doesn't seem all that bad to me?

**Comment by Piedone** [comment]
Since this operation only deletes a single record it indeed seems to be slow (but depending on your server configuration it can be normal as well, though this would need a very low-end machine especially for the DB).

Is this also slow with a vanilla Orchard instance? If no, then can it be that you app has some custom `IContentDefinitionEditorEvents` implementation or some code that listens to the `"ContentDefinitionManager"` signal and does something slow (these will run when such a delete happens)?

**Comment by Hazzamanic** [comment]
It deletes all the content items as well right? If there are a lot of content items that could cause the performance issues you are experiencing

**Comment by Piedone** [comment]
No, apart from the content type definition nothing else is deleted OOTB.

**Comment by byhub** [comment]
No we dont use  custom IContentDefinitionEditorEvents nor have any code listen to them.  What we have is huge number of content types. Would that be a reason?  if so it is an issue of more like an indexing or cache related delays? what is your opinion please?

**Comment by Piedone** [comment]
Hmm, that could indeed be an issue. I was mistaken, not just the `ContentTypeDefinitionRecord` instance but also the corresponding `ContentTypePartDefinitionRecord`s are deleted.

So adding indices there may help here, I'd encourage to test adding them on these columns and see what happens:

- `ContentTypeDefinitionRecord.Name` (used for looking up the content definition)
- `ContentTypePartDefinitionRecord.ContentTypeDefinitionRecord_Id (used to look up TypePartDefinitions corresponding to the content definition)

**Comment by Hazzamanic** [comment]
https://github.com/OrchardCMS/Orchard/blob/dev/src/Orchard.Web/Modules/Orchard.ContentTypes/Services/ContentDefinitionService.cs#L179 

Am I looking at totally the wrong code here? The value passed in is true from the admin controller.

**Comment by Piedone** [comment]
You're right, it indeed removes all content items, that's new for me. I was just looking at `ContentDefinitionManager`.

So indeed if you have many content items associated with that content type then several seconds is not out of ordinary.

**Comment by byhub** [comment]
Actually we dont have any content items associated with content types.. thats why we got confused.. it shouldnt be taking several seconds to delete 1 or 2 entities (which makes up contenttype together) from DB right?

**Comment by Piedone** [comment]
It shouldn't, check my previous comment: https://github.com/OrchardCMS/Orchard/issues/7828#issuecomment-325397881
